overlord/snapstate/snapstate.go

```diff
@@ -447,3 +447,39 @@ func GadgetInfo(s *state.State) (*snap.Info, error) {
 
 	return nil, state.ErrNoState
 }
+
+// MigrateToTypeInState migrates the state to have a snap type in the snap state of each setup snap.
+func MigrateToTypeInState(s *state.State, commit func()) error {
+	var stateMap map[string]*SnapState
+
+	s.Lock()
+	err := s.Get("snaps", &stateMap)
```

Comment:
@mvo5 by design these are run very early in loadState in overlord, this is before we have even created the managers or the state engine, so the only important locking is to try to update the state and patch-level in one go (or be incremental/idempotent otherwise), the rest needs is just to make the state methods happy while being sure not to deadlock